From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: VytskaLT <VytskaLT@protonmail.com>
Date: Sat, 17 Apr 2021 10:43:08 +0300
Subject: [PATCH] Cache block break animation packet

Original code by Electroid, licensed under GPL v3
You can find the original code on https://github.com/Electroid/SportPaper

diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 450c7a1033213b6515a2f8efa7ac7051bafb8cb3..b900324d9824b42cfa1af13e380fee13570e79ab 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -1830,6 +1830,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
         if (entity instanceof Player) entityhuman = (Player) entity;
         // CraftBukkit end
 
+        ClientboundBlockDestructionPacket packet = null; // SportPaper - Cache block break animation packet
         while (iterator.hasNext()) {
             ServerPlayer entityplayer = (ServerPlayer) iterator.next();
 
@@ -1845,7 +1846,10 @@ public class ServerLevel extends Level implements WorldGenLevel {
                 // CraftBukkit end
 
                 if (d0 * d0 + d1 * d1 + d2 * d2 < 1024.0D) {
-                    entityplayer.connection.send(new ClientboundBlockDestructionPacket(entityId, pos, progress));
+                    // SportPaper start
+                    if (packet == null) packet = new ClientboundBlockDestructionPacket(entityId, pos, progress);
+                    entityplayer.connection.send(packet);
+                    // SportPaper end
                 }
             }
         }
